kolibri/plugins/coach/assets/src/views/lessons/LessonSummaryPage/sidePanels/LessonResourceSelection/index.vue

```diff
@@ -0,0 +1,161 @@
+<template>
+
+  <SidePanelModal
+    alignment="right"
+    sidePanelWidth="700px"
+    closeButtonIconType="close"
+    @closePanel="closeSidePanel"
+    @shouldFocusFirstEl="() => null"
+  >
+    <template #header>
+      <div style="display: flex; gap: 8px; align-items: center">
+        <KIconButton
+          v-if="goBack"
+          icon="back"
+          @click="goBack"
+        />
+        <h1 class="side-panel-title">{{ view.title }}</h1>
+      </div>
+    </template>
+    <div v-if="loading">
+      <KCircularLoader />
+    </div>
+    <component
+      :is="view.component"
+      v-else-if="!view.guard || view.guard()"
+    />
+
+    <template #bottomNavigation>
+      <div class="bottom-nav-container">
+        <KButton
+          primary
+          :text="saveAndFinishAction$()"
+          @click="closeSidePanel"
+        />
+      </div>
+    </template>
+  </SidePanelModal>
+
+</template>
+
+
+<script>
+
+  import SidePanelModal from 'kolibri-common/components/SidePanelModal';
+  import { coreStrings } from 'kolibri/uiText/commonCoreStrings';
+  import { PageNames } from '../../../../../constants';
+  import SelectionIndex from './subPages/SelectionIndex';
+  import useResourceSelection from './useResourceSelection';
+  import SelectFromBookmarks from './subPages/SelectFromBookmarks.vue';
+  import SelectFromChannels from './subPages/SelectFromChannels.vue';
+  import { ResourceSelectionView } from './constants';
+
+  export default {
+    name: 'LessonResourceSelection',
+    components: {
+      SidePanelModal,
+      SelectionIndex,
+      SelectFromBookmarks,
+      SelectFromChannels,
+    },
+    setup() {
+      const { loading } = useResourceSelection();
+
+      const { saveAndFinishAction$, selectFromBookmarks$ } = coreStrings;
+
+      return {
+        loading,
+        selectFromBookmarks$,
+        saveAndFinishAction$,
+      };
+    },
+    computed: {
+      topicId() {
+        const { topicId } = this.$route.query;
+        return topicId;
+      },
+      viewId() {
+        const { viewId } = this.$route.params;
+        if (Object.values(ResourceSelectionView).includes(viewId)) {
+          return viewId;
+        }
+        return ResourceSelectionView.SELECTION_INDEX;
+      },
+      view() {
+        const componentMap = {
+          [ResourceSelectionView.SELECTION_INDEX]: {
+            title: this.$tr('manageLessonResourcesTitle'),
+            component: SelectionIndex,
+          },
+          [ResourceSelectionView.SELECT_FROM_BOOKMARKS]: {
+            title: this.selectFromBookmarks$(),
+            component: SelectFromBookmarks,
+            back: ResourceSelectionView.SELECTION_INDEX,
+          },
+          [ResourceSelectionView.SELECT_FROM_CHANNELS]: {
+            title: this.$tr('manageLessonResourcesTitle'),
+            component: SelectFromChannels,
+            back: ResourceSelectionView.SELECTION_INDEX,
+            guard: () => !!this.topicId,
+          },
```

Comment:
Not a blocking comment, just want to share a thought I've been having around this kind of pattern that is relatively common in our code.

These objects are structured very much alike to a VueRouter route which makes me wonder if we should consider using VueRouter to handle this behavior instead. Mostly this comes from my sense that this is a job that the VueRouter is purpose-built to handle.

That all said, my meager attempts at doing this myself have run into issues so just something to think about, particularly as we move toward redesigning the SidePanelModal itself.

Reply:
Yees, thats true, I also thought about that, my first thought was that it was gonna be easier to have the business logic to manage the guards or the back page inside the component rather than doing it in the routes array. But I am gonna give it a try and actually see the implications of moving this as route children.